aspnetcore/blazor/Security-guidance.md

```diff
@@ -0,0 +1,254 @@
+  * Take into account that static and instance methods can be exposed to JavaScript clients and that you should take special considerations to avoid sharing state between sessions.
+    * For instance methods exposed through the use of DotNetReference objects, if originally created through DI should be registered as scoped objects (this applies to any DI service your blazor server-side app uses).
+    * For static methods you should avoid having any sort of state that can't be scoped to the client.
+  * Avoid using user input for JS interop calls or validate the input if you have to to avoid XSS attacks. Consider also using CSP to disable 'eval' and other unsafe JS primitives.
+* We recommend that you avoid implementing your own dispatching of .NET invocations on top of our implementation. While its doable to implement a more general .NET method dispatching infrastructure on top of the one provided by the framework, we strongly discourage that you do so, as exposing .NET methods to the browser should be an explicit action that needs the same careful thought as exposing any other API member.
+
+### Consuming events
+Events are another entry point to your application, so the same rules as any other endpoint apply for consuming events. In particular, a bad behaving client can send whatever fake data they want as the payload for the event.
+
+For example, a change event for a select could send a value that is not within the options, or a textbox could send any text data to the server bypassing any client side validation.
+
+We recommend that you validate the input for any event that your app handles. The forms components that we provide as part of the framework already perform those validations, but if you are writing your own components you need to validate the inputs as appropriate.
+
+In blazor server-side events are asynchronous by nature, so multiple events can be dispatched to the server before the application has time to react by producing a new render. This has some security implications we need to be aware of.
+
+Limiting client actions on the application needs to be done inside event handlers and not depend on the current rendered view state. 
+
+For example, imagine a counter where we want to only allow users to click the counter 3 times. If the logic for our counter is similar to this
+```
+<p>@count<p>
+@if(count < 3)
+{
+  <button @onclick="IncrementCount" value="Increment count" />
+}
+
+@code 
+{
+  int count = 0;
+  void IncrementCount() => count++;
+}
+```
+a client can dispatch one or more increment events before we produce a new render and result in the count being incremented over 3. The right way to achieve something like this is as follows.
+```
+<p>Count @count<p>
+@if(count < 3)
+{
+  <button @onclick="IncrementCount" value="Increment count" />
+}
+
+@code 
+{
+  int count = 0;
+  void IncrementCount() => count < 3 ? count++ : count;
```

Comment:
To avoid confusing people less familiar with C#:

```suggestion
  void IncrementCount()
  {
      if (count < 3)
      {
          count++;
      }
  }
```